### DISCUSSION TITLE
Best method to manually poll for jobs?

### DISCUSSION THREAD

**Discussion by prescience-data** [taskforcesh/bullmq#462] [Q&A] [answered]
I've just tried to implement a manual job getter, but it seems to return `void` if not jobs are immediately on the queue.

Before going further down this path, I wanted to check if this is a rational method to poll for jobs, or if there is a better way?

Without any real benchmarking my concern is that this approach across hundreds of workers seems like it would be thrashing fairly hard with a sufficiently small `delay`...  🤔

(**Note:** The reason for requesting jobs manually is to allow workers to return a completed job result, but not accept a new job until a secondary process is triggered.)

```typescript
/**
 * Loops on a heartbeat until a job lock is acquired.
 *
 * @param {BullWorker} worker
 * @param {string} uuid
 * @param {number} delay
 * @return {Promise<SocketRequestJob>}
 */
export const acquireJob = async (
  worker: BullWorker,
  { uuid, delay }: JobAcquisitionOptions
): Promise<SocketRequestJob> => {
  // Set up the local variable to receive inbound jobs.
  let job: SocketRequestJob | undefined
  // Loop until a job lock is acquired.
  while (job === undefined) {
    job = (await worker.getNextJob(uuid)) || undefined
    // If no job was acquired, delay for a short delay before trying again.
    if (job === undefined) {
      await waitForTimeout(delay)
    }
  }
  // Return acquired job.
  return job
}


```

**Comment by ccollie** [comment]
I'd have to devote a few brain cells, but my approach would be to leverage events:

https://docs.bullmq.io/guide/events

See the following test: https://github.com/taskforcesh/bullmq/blob/297ec30086bd0d95b10718c128409e535414180f/src/test/test_events.ts#L29

**Reply by prescience-data** [reply]
Ah interesting, I hadn't considered the on add event ⚡🧠

**Comment by manast** [comment, marked as answer]
You do not need a delay, since  "getNextJob" is already a blocking operation that waits "drainDelay" before it expires. The default value for drainDelay is 5 seconds (note it is expressed in seconds not milliseconds). You can increase this value to reduce the loop interval. "getNextJob" is the most efficient way to get jobs from the queue,  I do not recommend using events for this usecase.

**Reply by prescience-data** [reply]
@manast  
> I do not recommend using events for this usecase.

Re the events, is that because it would trigger all the workers at the exact ms? or some other reason?

> blocking operation that waits "drainDelay" before it expires

So "drainDelay" refers to the worker draining the socket / stream when it finishes checking for any jobs and begins to disconnect? (in this case with an empty result?)

**Reply by manast** [reply]
> So "drainDelay" refers to the worker draining the socket / stream when it finishes checking for any jobs and begins to disconnect? (in this case with an empty result?)

No. When the queue is drained, i.e. does not have any items in the wait list, we issue a blocking command to redis to wait for a job to arrive. This command (brpoplpush) has a "timeout" parameter which specifies for how long the call should be blocking until it just returns, thats the "drainDelay", probably a better name would be more suitable here. The way redis works is that for all the clients that are blocked it will pick one in round robin fashion. If you have many workers you can increase this value considerably and mostly remove any overhead arising from the polling.

**Reply by prescience-data** [reply]
Ok so to put that in practical terms, this is the ideal implementation?
```typescript
export const acquireJob = async (
  worker: BullWorker,
  { uuid }: JobAcquisitionOptions
): Promise<SocketRequestJob> => {
  let job: SocketRequestJob | undefined
  while (job === undefined) {
    job = (await worker.getNextJob(uuid)) || undefined
    if (job) {
      return job
    }
  }
}
```

**Reply by manast** [reply]
I cannot say if it is ideal for your particular use case, but why would you have a loop inside acquireJob instead of moving the looping higher up?

**Reply by prescience-data** [reply]
That was just to make it simpler to describe for the example, in the real code that would be inline. 
But it sounds like it does what I think it does, which is excellent 🕺✌️

**Comment by prescience-data** [comment]
@manast 
I think I must be doing something wrong here.

Both the server and the worker (appear to be) connected to the `"flex"` queue, but the worker never picks up a job and Taskforce.sh is showing zero workers...

Logs:
![image](https://user-images.githubusercontent.com/65471523/114179061-dea33180-9981-11eb-8f8c-61aa5458d8c9.png)
Taskforce:
![image](https://user-images.githubusercontent.com/65471523/114179169-01354a80-9982-11eb-902b-35cef3a5158c.png)

**Reply by manast** [reply]
hmm, looks as if the workers are not really connected to the correct redis instance. I recommend you that you write a minimal POC in one file where you both add the jobs and fetch them manually to verify everything works as expected.

**Reply by prescience-data** [reply]
I just realised I was passing the worker a `prefix` option and the server (queue dispatcher) no prefix 🙈

**Comment by prescience-data** [comment]
Just to close this out for anyone stumbling across this in the future, I misunderstood @manast explanation around the `drainDelay` - I thought a shorter delay was required, but you actually want a longer delay.

The delay isn't related to a "polling" cycle, it refers to how long the worker stays connected to the queue waiting for a job before giving up. So a long number = waiting for a job for a long time (which is what I wanted in this situation).
